File: data/src/main/java/com/marvelcomics/brito/data/repository/character/CharacterRepository.java
package com.marvelcomics.brito.data.repository.character;

import com.marvelcomics.brito.data.datasource.remote.response.CharacterResponse;
import com.marvelcomics.brito.data.datasource.remote.response.model.RemoteMarvelContainer;

import io.reactivex.Observable;

public interface CharacterRepository {

    Observable<RemoteMarvelContainer<CharacterResponse>> characters(String name);
}

File: data/src/main/java/com/marvelcomics/brito/data/repository/characters/CharacterRepository.kt
package com.marvelcomics.brito.data.repository.characters

import com.marvelcomics.brito.data.datasource.remote.mapper.CharacterMapper
import com.marvelcomics.brito.data.webservice.MarvelWebService
import com.marvelcomics.brito.domain.entity.CharacterEntity
import com.marvelcomics.brito.domain.repository.ICharacterRepository
import com.marvelcomics.brito.infrastructure.exception.MarvelMapperException

class CharacterRepository(
    private val webService: MarvelWebService,
    private val characterMapper: CharacterMapper
) : ICharacterRepository {
    override suspend fun getCharacters(name: String): List<CharacterEntity> {
        return characterMapper.transform(webService.characters(name))
            ?: run { throw MarvelMapperException("Error mapping character", null) }
    }
}

